jedi/plugins/django.py

```diff
@@ -109,25 +131,160 @@ def _create_manager_for(cls, manager_cls='BaseManager'):
     return None
 
 
-def _new_dict_filter(cls):
-    filters = cls.get_filters(is_instance=True, include_metaclasses=False)
+def _new_dict_filter(cls, is_instance):
+    filters = list(cls.get_filters(
+        is_instance=is_instance,
+        include_metaclasses=False,
+        include_type_when_class=False)
+    )
     dct = {
-        name.string_name: DjangoModelName(cls, name)
-        for filter_ in reversed(list(filters))
+        name.string_name: DjangoModelName(cls, name, is_instance)
+        for filter_ in reversed(filters)
         for name in filter_.values()
     }
-    manager = _create_manager_for(cls)
-    if manager:
-        dct['objects'] = manager.name
+    if is_instance:
+        # Replace the objects with a name that amounts to nothing when accessed
+        # in an instance. This is not perfect and still completes "objects" in
+        # that case, but it at least not inferes stuff like `.objects.filter`.
+        # It would be nicer to do that in a better way, so that it also doesn't
+        # show up in completions, but it's probably just not worth doing that
+        # for the extra amount of work.
+        dct['objects'] = EmptyCompiledName(cls.inference_state, 'objects')
+
     return DictFilter(dct)
 
 
+def is_django_model_base(value):
+    return value.py__name__() == 'ModelBase' \
+        and value.get_root_context().py__name__() == 'django.db.models.base'
+
+
 def get_metaclass_filters(func):
-    def wrapper(cls, metaclasses):
+    def wrapper(cls, metaclasses, is_instance):
         for metaclass in metaclasses:
-            if metaclass.py__name__() == 'ModelBase' \
-                    and metaclass.get_root_context().py__name__() == 'django.db.models.base':
-                return [_new_dict_filter(cls)]
+            if is_django_model_base(metaclass):
+                return [_new_dict_filter(cls, is_instance)]
+
+        return func(cls, metaclasses, is_instance)
+    return wrapper
+
 
-        return func(cls, metaclasses)
+def tree_name_to_values(func):
+    def wrapper(inference_state, context, tree_name):
+        result = func(inference_state, context, tree_name)
+        if tree_name.value in ('create', 'filter', 'exclude', 'update', 'get',
+                               'get_or_create', 'update_or_create'):
```

Comment:
It took me a while to work out what the common theme was for these methods. Would it be worth pulling them out to a named constant perhaps or adding a comment here about the intent of this branch?

Maybe `FIELD_NAMES_AS_KWARGS_METHODS` or `SIGNATURE_MATCHES_MODEL_METHODS`?

Reply:
Good point. Will do and then merge! :)

I tend to not comment enough anyway since for a long time I was the only one editing.